extension/proxmox: add console access for instances

shwstppr opened the pull request:
### Description

This PR introduces console access support for instances deployed using Orchestrator Extensions, available via either VNC or a direct URL.

- CloudStack queries the extension using the getconsole action.
- For VNC-based access, the extension must return host/port/ticket details. CloudStack then forwards these to the Console Proxy VM (CPVM) in the instance’s zone. It is assumed that the CPVM can reach the specified host and port.
- For direct URL access, the extension returns a console URL with the protocol set to `direct`. The URL is then provided directly to the user.
- The built-in Proxmox Orchestrator Extension now supports console access via VNC. The extension calls the Proxmox API to fetch console details and returns them in the required format.

Also, adds changes to send caller details to the extension payload.
```
# cat /var/lib/cloudstack/management/extensions/Proxmox/02b650f6-bb98-49cb-8cac-82b7a78f43a2.json | jq
{
  "caller": {
    "roleid": "6b86674b-7e61-11f0-ba77-1e00c8000158",
    "rolename": "Root Admin",
    "name": "admin",
    "roletype": "Admin",
    "id": "93567ed9-7e61-11f0-ba77-1e00c8000158",
    "type": "ADMIN"
  },
  "virtualmachineid": "126f4562-1f0f-4313-875e-6150cabeb72f",
  ...
```

Documentation PR: https://github.com/apache/cloudstack-documentation/pull/560

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### Screenshots (if appropriate):

https://github.com/user-attachments/assets/3c368722-801b-4aba-979c-4bfded4676d0


### How Has This Been Tested?

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->

#### How did you try to break this feature and the system with this change?

<!-- see how your change affects other areas of the code, etc. -->

<!-- Please read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/main/CONTRIBUTING.md) document -->

shwstppr commented:
@blueorangutan pacakge

shwstppr commented:
@blueorangutan package

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ el10 ✔️ debian ✔️ suse15. SL-JID 15078

shwstppr commented:
@blueorangutan package

shwstppr commented:
@blueorangutan test

blueorangutan commented:
@shwstppr a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-14403)</b>
Environment: kvm-ol8 (x2), zone: Advanced Networking with Mgmt server ol8
Total time taken: 58393 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr11601-t14403-kvm-ol8.zip
Smoke tests completed. 146 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_03_deploy_and_scale_kubernetes_cluster | `Failure` | 1.17 | test_kubernetes_clusters.py
test_04_autoscale_kubernetes_cluster | `Failure` | 27.94 | test_kubernetes_clusters.py

kiranchavala commented:
@blueorangutan test

kiranchavala reviewed (approved):
LGTM

| Test Case Execution                                         | Result |
|-------------------------------------------------------------|--------|
| Test the console access of a proxmox vm   from the UI       | Pass   |
| Test the listConsoleSessions api   call                     | Pass   |
| Test the api createConsoleEndpoint                          | Pass   |
| Test the Direct Url based console                           | Pass   |
| Test from multiple management servers   enabled environment | Pass   |
| Test from a ssl enabled   cloudstack-management server      | Pass   |

blueorangutan commented:
<b>[SF] Trillian test result (tid-14431)</b>
Environment: kvm-ol8 (x2), zone: Advanced Networking with Mgmt server ol8
Total time taken: 51538 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr11601-t14431-kvm-ol8.zip
Smoke tests completed. 147 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

rohityadavcloud reviewed (changes requested):
Changes requested:

1. Improve spec & documentation on getconsole access types, vnc and url. This does not make sense to me: " the extension returns a console URL with the protocol set to direct, url, or link. The URL is then provided directly to the user." Please also update https://github.com/apache/cloudstack-documentation/pull/560/files

2. In the getconsole, can we pass some kind of account or role type information, which can be used by Extension authors to return different URLs for different types of roles. For example, return an internal URL if the caller is a root admin versus a normal/public URL if the caller is a public tenant.

shwstppr commented:
@blueorangutan package

shwstppr commented:
@rohityadavcloud @harikrishna-patnala addressed comments. Now all extension payloads will contain `caller` field in `externaldetails` which will give basic details of the caller. Please see if that works for you

```
# cat /var/lib/cloudstack/management/extensions/Proxmox/5811778e-bded-4576-96bb-3d529d8c0a33.json | jq
{
  "virtualmachineid": "126f4562-1f0f-4313-875e-6150cabeb72f",
  "cloudstack.vm.details": {
    "id": 45,
    "name": "i-2-45-VM",
    "state": "Running",
    "type": "User",
    "cpus": 1,
    "minSpeed": 500,
    "maxSpeed": 500,
    "minRam": 536870912,
    "maxRam": 536870912,
    "arch": "x86_64",
    "bootArgs": "",
    "enableHA": false,
    "limitCpuUse": false,
    "enableDynamicallyScaleVm": false,
    "vncPassword": "uHQHQ4niy8f4ogwCM1k-rw",
    "details": {
      "proxmox_vmid": "103",
      "cpuOvercommitRatio": "2.0",
      "External:template_type": "template",
      "External:template_id": "101",
      "Message.ReservedCapacityFreed.Flag": "false"
    },
    "uuid": "126f4562-1f0f-4313-875e-6150cabeb72f",
    "enterHardwareSetup": false,
    "disks": [],
    "nics": [],
    "configDriveLocation": "SECONDARY",
    "guestOsDetails": {},
    "extraConfig": {},
    "networkIdToNetworkNameMap": {}
  },
  "virtualmachinename": "i-2-45-VM",
  "externaldetails": {
    "caller": {
      "roleid": "6b86674b-7e61-11f0-ba77-1e00c8000158",
      "rolename": "Root Admin",
      "name": "admin",
      "roletype": "Admin",
      "id": "93567ed9-7e61-11f0-ba77-1e00c8000158",
      "type": "ADMIN"
    },
    "host": {
      "node": "ref-trl-9283-k-mr8-abhishek-kumar-proxmox1",
      "network_bridge": "cloudbr1",
      "secret": "c0541327-d325-4a0f-8e14-3baf9eeb2da0",
      "user": "root@pam",
      "verify_tls_certificate": "false",
      "url": "10.0.35.44",
      "token": "token1"
    },
    "virtualmachine": {
      "template_type": "template",
      "template_id": "101"
    }
  }
}
```

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9  ✔️ debian ✖️ suse15. SL-JID 15161

shwstppr commented:
@blueorangutan test ol8 vmware-80u3

blueorangutan commented:
@shwstppr a [SL] Trillian-Jenkins test job (ol8 mgmt + vmware-80u3) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-14439)</b>
Environment: vmware-80u3 (x2), zone: Advanced Networking with Mgmt server ol8
Total time taken: 78614 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr11601-t14439-vmware-80u3.zip
Smoke tests completed. 144 look OK, 3 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_08_upgrade_kubernetes_ha_cluster | `Failure` | 3681.92 | test_kubernetes_clusters.py
test_01_ssl_offloading_isolated_network | `Failure` | 895.58 | test_ssl_offloading.py
test_02_ssl_offloading_project_vpc | `Failure` | 764.59 | test_ssl_offloading.py
test_01_vpn_usage | `Error` | 1.11 | test_usage.py

shwstppr commented:
Some VMware test failures unrelated. Similar results seen here in healthcheck, https://github.com/apache/cloudstack/pull/11523#issuecomment-3315974820
Merging based on reviews and test results
